src/main/java/org/jboss/pnc/reqour/rest/endpoints/CloneEndpointImpl.java

```diff
@@ -0,0 +1,108 @@
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.pnc.reqour.rest.endpoints;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import jakarta.ws.rs.core.Response;
+import lombok.extern.slf4j.Slf4j;
+import org.eclipse.microprofile.context.ManagedExecutor;
+import org.jboss.pnc.api.reqour.dto.Callback;
+import org.jboss.pnc.api.reqour.dto.RepositoryCloneRequest;
+import org.jboss.pnc.api.reqour.dto.RepositoryCloneResponseCallback;
+import org.jboss.pnc.reqour.api.CloneEndpoint;
+import org.jboss.pnc.reqour.common.callbacksender.CallbackSender;
+import org.jboss.pnc.reqour.common.exceptions.GitException;
+import org.jboss.pnc.reqour.common.taskid.TaskIDGenerator;
+import org.jboss.pnc.reqour.common.taskid.TaskIDMapper;
+import org.jboss.pnc.reqour.facade.api.CloneProvider;
+import org.jboss.pnc.reqour.facade.clone.CloneProviderPicker;
+
+import java.util.UUID;
+
+@ApplicationScoped
+@Slf4j
+public class CloneEndpointImpl implements CloneEndpoint {
+
+    private final CloneProviderPicker delegatePicker;
+    private final ManagedExecutor executor;
+    private final CallbackSender callbackSender;
+    private final TaskIDGenerator<UUID> taskIDGenerator;
+    private final TaskIDMapper<UUID> taskIDMapper;
+
+    @Inject
+    public CloneEndpointImpl(
+            CloneProviderPicker delegatePicker,
+            ManagedExecutor executor,
+            CallbackSender callbackSender,
+            TaskIDGenerator<UUID> taskIDGenerator,
+            TaskIDMapper<UUID> taskIDMapper) {
+        this.delegatePicker = delegatePicker;
+        this.executor = executor;
+        this.callbackSender = callbackSender;
+        this.taskIDGenerator = taskIDGenerator;
+        this.taskIDMapper = taskIDMapper;
+    }
+
+    @Override
+    public void clone(RepositoryCloneRequest cloneRequest) {
+        var taskID = taskIDGenerator.generateTaskID();
+
+        // Picking of the correct delegate (based on request) is better to be done in the current thread
+        // In order to fail in case the cloneRequest.getType() is invalid
+        CloneProvider cloneProvider = delegatePicker.pickProvider(cloneRequest.getType());
+
+        executor.runAsync(() -> {
+            log.info("Starting the task with ID={}", taskID);
+            cloneProvider.clone(cloneRequest);
+        }).handle((_res, ex) -> fireCallback(cloneRequest, ex, taskID));
```

Comment:
as we were yesterday discussing in person, just remembered: this will be more elegant to change to `.whenComplete()`, hence, do not be forced to `return null;` from `fireCallback`

Reply:
rewritten this way ^